### PR TITLE
OHRM5X-1640: Custom LangPack Bug Fix

### PR DESCRIPTION
**Pull request by ManulMax**

### CODE REVIEW

File: src/client/src/orangehrmAdminPlugin/pages/languageTranslation/languageTranslation.vue [hunk 1]
```diff
@@ -19,7 +19,12 @@
  -->
 
 <template>
-  <div class="orangehrm-background-container">
+  <oxd-alert
```
Comment (Super-Chama) [inline src/client/src/orangehrmAdminPlugin/pages/languageTranslation/languageTranslation.vue:22]: @ManulMax what's this requirement? we normally show no records found using a toast message right?

File: installer/Migration/V5_2_0/Migration.php [hunk 2]
```diff
@@ -62,6 +62,18 @@ public function up(): void
...
+            ->delete('ohrm_i18n_group')
+            ->andWhere('ohrm_i18n_group.name = :name')
+            ->setParameter('name', 'directory')
+            ->executeQuery();
+
+        $this->createQueryBuilder()
+            ->delete('ohrm_i18n_group')
+            ->andWhere('ohrm_i18n_group.name = :name')
+            ->setParameter('name', 'branding')
+            ->executeQuery();
```
Comment (RajithaKumara) [inline installer/Migration/V5_2_0/Migration.php:76]:
```php
        $qb = $this->createQueryBuilder()
            ->delete('ohrm_i18n_group');
        $qb->where($qb->expr()->in('ohrm_i18n_group.name', ':groups'))
            ->setParameter('groups', ['directory', 'branding'], \Doctrine\DBAL\Connection::PARAM_STR_ARRAY)
            ->executeQuery();
```

File: src/plugins/orangehrmAdminPlugin/Controller/LanguageTranslationController.php [hunk 3]
```diff
@@ -38,9 +38,11 @@ public function preRender(Request $request): void
             $language = $this->getLocalizationService()->getLocalizationDao()
                 ->getLanguageById($languageId);
             $languagePackage = $language->getName();
+            $isAdded = $language->isAdded();
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Controller/LanguageTranslationController.php:41]:
```suggestion
            if ($language->isAdded() && $language->isEnabled()) {
                throw new RequestForwardableException(NoRecordsFoundController::class . '::handle');
            }
```

File: src/plugins/orangehrmAdminPlugin/Controller/LanguageTranslationController.php [hunk 4]
```diff
@@ -38,9 +38,11 @@ public function preRender(Request $request): void
...
             $sourceLanguage = 'English (United States)';
             $component->addProp(new Prop('language-package', Prop::TYPE_STRING, $languagePackage));
             $component->addProp(new Prop('source-language', Prop::TYPE_STRING, $sourceLanguage));
+            $component->addProp(new Prop('is-added', Prop::TYPE_BOOLEAN, $isAdded));
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Controller/LanguageTranslationController.php:45]:
```suggestion
```

File: src/client/src/orangehrmAdminPlugin/pages/languageTranslation/languageTranslation.vue [hunk 5]
```diff
@@ -172,6 +180,10 @@ export default {
       type: String,
       required: true,
     },
+    isAdded: {
+      type: Boolean,
+      required: true,
+    },
```
Comment (RajithaKumara) [inline src/client/src/orangehrmAdminPlugin/pages/languageTranslation/languageTranslation.vue:186]:
```suggestion
```

File: src/client/src/orangehrmAdminPlugin/pages/languageTranslation/languageTranslation.vue [hunk 6]
```diff
@@ -19,7 +19,12 @@
  -->
 
 <template>
-  <div class="orangehrm-background-container">
+  <oxd-alert
+    :show="!isAdded"
+    type="warn"
+    :message="$t('general.no_records_found')"
+  ></oxd-alert>
```
Comment (RajithaKumara) [inline src/client/src/orangehrmAdminPlugin/pages/languageTranslation/languageTranslation.vue:26]:
```suggestion
```

File: src/client/src/orangehrmAdminPlugin/pages/languageTranslation/languageTranslation.vue [hunk 7]
```diff
@@ -19,7 +19,12 @@
...
+  <div v-if="isAdded" class="orangehrm-background-container">
```
Comment (RajithaKumara) [inline src/client/src/orangehrmAdminPlugin/pages/languageTranslation/languageTranslation.vue:27]:
```suggestion
  <div class="orangehrm-background-container">
```

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmDashboardPlugin/Api/EmployeeOnLeaveAPI.php [hunk 8]
```diff
@@ -32,18 +32,18 @@
...
 class EmployeeOnLeaveAPI extends Endpoint implements CollectionEndpoint
 {
     use DateTimeHelperTrait;
-    use LeavePeriodServiceTrait;
+    use LeaveConfigServiceTrait;
     use EmployeeOnLeaveServiceTrait;
 
     public const PARAMETER_DATE = 'date';
-    public const PARAMETER_LEAVE_PERIOD = 'leavePeriod';
+    public const META_PARAMETER_LEAVE_PERIOD_DEFINED =  'leavePeriod';
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Api/EmployeeOnLeaveAPI.php:46]:
```suggestion
    public const META_PARAMETER_LEAVE_PERIOD_DEFINED =  'leavePeriodDefined';
```

File: src/plugins/orangehrmDashboardPlugin/Service/ChartService.php [hunk 9]
```diff
@@ -73,39 +73,28 @@ static function (SubunitCountPair $x, SubunitCountPair $y) {
...
         $locationEmployeeCount = $this->getChartDao()->getEmployeeDistributionByLocation();
-        $unassignedEmployeeCount =  $this->getLocationUnassignedEmployeeCount($locationEmployeeCount);
 
-        $otherArray = [];
-        if (count($locationEmployeeCount) > $limit) {
-            $otherArray = array_slice($locationEmployeeCount, $limit);
-            $locationEmployeeCount = array_slice($locationEmployeeCount, 0, $limit);
+        $otherCount = 0;
+        $assignedEmployeeCount = 0;
+        foreach ($locationEmployeeCount as $key => $employeeCount) {
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Service/ChartService.php:86]:
```suggestion
        foreach ($locationEmployeeCount as $index => $employeeCount) {
```
refactor variable name

File: src/plugins/orangehrmDashboardPlugin/Service/ChartService.php [hunk 10]
```diff
@@ -73,39 +73,28 @@ static function (SubunitCountPair $x, SubunitCountPair $y) {
...
+            $assignedEmployeeCount += $employeeCount->getEmployeeCount();
+            if ($key >= $limit) {
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Service/ChartService.php:88]:
```suggestion
            if ($index >= $limit + 1) {
```
refactor carefully

File: src/plugins/orangehrmDashboardPlugin/Service/ChartService.php [hunk 11]
```diff
@@ -73,39 +73,28 @@ static function (SubunitCountPair $x, SubunitCountPair $y) {
...
+            }
         }
 
-        $otherCount = 0;
-        foreach ($otherArray as $locationCountPair) {
-            $otherCount += $locationCountPair->getEmployeeCount();
+        $totalActiveEmployee = $this->getChartDao()->getTotalActiveEmployeeCount();
+        $unassignedEmployeeCount = $totalActiveEmployee - $assignedEmployeeCount;
+
+        if (count($locationEmployeeCount) > $limit) {
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Service/ChartService.php:96]:
```suggestion
        if (count($locationEmployeeCount) > $limit + 1) {
```
refactor carefully and check unit tests. if you can't add multiple service unit tests. check code coverage after these bug fixes

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/client/src/orangehrmAdminPlugin/pages/languageTranslation/languageTranslation.vue (hunk 1 above)
Comment (ManulMax) [reply src/client/src/orangehrmAdminPlugin/pages/languageTranslation/languageTranslation.vue]: Changed it

File: installer/Migration/V5_2_0/Migration.php (hunk 2 above)
Comment (ManulMax) [reply installer/Migration/V5_2_0/Migration.php]: Add changes

File: src/plugins/orangehrmAdminPlugin/Controller/LanguageTranslationController.php (hunk 4 above)
Comment (ManulMax) [reply src/plugins/orangehrmAdminPlugin/Controller/LanguageTranslationController.php]: Add changes

File: src/client/src/orangehrmAdminPlugin/pages/languageTranslation/languageTranslation.vue (hunk 5 above)
Comment (ManulMax) [reply src/client/src/orangehrmAdminPlugin/pages/languageTranslation/languageTranslation.vue]: Add changes

File: src/client/src/orangehrmAdminPlugin/pages/languageTranslation/languageTranslation.vue (hunk 6 above)
Comment (ManulMax) [reply src/client/src/orangehrmAdminPlugin/pages/languageTranslation/languageTranslation.vue]: Add changes

File: src/plugins/orangehrmDashboardPlugin/Api/EmployeeOnLeaveAPI.php (hunk 8 above)
Comment (ManulMax) [reply src/plugins/orangehrmDashboardPlugin/Api/EmployeeOnLeaveAPI.php]: Add changes

File: installer/Migration/V5_2_0/Migration.php [hunk 12]
```diff
@@ -62,6 +62,14 @@ public function up(): void
         $this->updateOrganizationStructure();
         $this->updateHomePage('Admin', 'dashboard/index');
         $this->updateHomePage('ESS', 'dashboard/index');
+
+        $qb = $this->createQueryBuilder()->delete('ohrm_i18n_group');
+        $qb->where($qb->expr()->in('ohrm_i18n_group.name', ':groups'))
+            ->setParameter(
+                'groups',
+                ['directory', 'branding'],
+                \Doctrine\DBAL\Connection::PARAM_STR_ARRAY
```
Comment (RajithaKumara) [inline installer/Migration/V5_2_0/Migration.php:71]: format the code using IDE with optimize import tick

**Review by RajithaKumara — APPROVED** [review]

**Review by Super-Chama — APPROVED** [review]

File: installer/Migration/V5_2_0/Migration.php (hunk 12 above)
Comment (RajithaKumara) [reply installer/Migration/V5_2_0/Migration.php:92]: if you formatted this correctly; should like below
```php
    $qb->where($qb->expr()->in('ohrm_i18n_group.name', ':groups'))
            ->setParameter(
                'groups',
                ['directory', 'branding'],
                Connection::PARAM_STR_ARRAY
            )->executeQuery();
```

File: src/plugins/orangehrmDashboardPlugin/Api/Model/BulkEmployeesOnLeaveModel.php [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+namespace OrangeHRM\Dashboard\Api\Model;
+
+use OrangeHRM\Core\Api\V2\Serializer\CollectionNormalizable;
+use OrangeHRM\Core\Api\V2\Serializer\ModelConstructorArgsAwareInterface;
+use OrangeHRM\Core\Traits\UserRoleManagerTrait;
+use OrangeHRM\Entity\Employee;
+use OrangeHRM\Entity\Leave;
+
+class BulkEmployeesOnLeaveModel implements CollectionNormalizable, ModelConstructorArgsAwareInterface
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Api/Model/BulkEmployeesOnLeaveModel.php:28]:
```suggestion
class EmployeesOnLeaveListModel implements CollectionNormalizable, ModelConstructorArgsAwareInterface
```
refactor class name to `EmployeesOnLeaveListModel`

File: src/plugins/orangehrmDashboardPlugin/Api/Model/BulkEmployeesOnLeaveModel.php [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+                    'middleName' => $leave->getEmployee()->getMiddleName(),
+                    'employeeId' => $leave->getEmployee()->getEmployeeId(),
+                    'terminationId' => $leave->getEmployee()->getEmployeeTerminationRecord() ? $leave->getEmployee()
+                        ->getEmployeeTerminationRecord()->getId() : null
+                ],
+                'duration' => $leave->getDecorator()->getLeaveDuration(),
+                'endTime' => $leave->getDecorator()->getEndTime(),
+                'startTime' => $leave->getDecorator()->getStartTime(),
+            ];
+            if ($this->getUserRoleManager()
```
Comment (RajithaKumara) [inline src/plugins/orangehrmDashboardPlugin/Api/Model/BulkEmployeesOnLeaveModel.php:66]: this should be covered in endpoint integration tests. 
you have to change `userId` and change loggedin user while running each test

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

**Comment by RajithaKumara** [comment]
Moved #1442

**Pull request closed without merging**